Documentation/design-docs/first-class-structs.md

```diff
@@ -15,6 +15,261 @@ Primary Objectives
 Secondary Objectives
 * No “swizzling” or lying about struct types – they are always struct types
  - No confusing use of GT_LCL_FLD to refer to the entire struct as a different type
+  
+Normalizing Struct Types
+------------------------
+We would like to facilitate full enregistration of structs with the following properties:
+1. Its fields are infrequently accessed, and
```

Comment:
I'm skeptical that enregistering structs only when fields are infrequently accessed is going to provide a lot of value. If fields aren't accessed then it all mostly resumes to struct copying. So that
```C#
x = f.a;
y = f.a;
```
with CSE and enregistration would hopefully generate just
```asm
mov eax, [rsi+8]
mov ebx, eax
```
Nice but I'm not sure how common such scenarios are.
Ultimately we'll probably need to consider the possibility of `LCL_FLD` working on enregistered struct by generating appropriate shift and masking code.

Reply:
Ah, right, that case is really ugly today. Speaking of struct return, I also have concerns about the general case where the struct is returned via a buffer and this transform is done pretty early in phase ordering. My initial impression is that it would be better to postpone that to lowering.